src/cipher.rs

```diff
@@ -0,0 +1,82 @@
+use crate::IV_SIZE;
+use aes::{Aes128, Aes256};
+use ctr::{
+    stream_cipher::{NewStreamCipher, SyncStreamCipher},
+    Ctr128,
+};
+use generic_array::GenericArray;
+use zeroize::Zeroize;
+
+/// Stream ciphers (i.e. AES-CTR) used in AES-SIV constructions
+// TODO: replace this with traits from the `stream-cipher` crate?
+pub trait Cipher {
+    /// Create a new CTR instance
+    fn new(key: &[u8]) -> Self;
```

Comment:
Does the `stream-cipher` crate have a way of more gracefully failing the init of a Cipher if, i.e. not the right key length was provided?

Reply:
Haha this whole trait is bogus, although this is "ok" in as much as this is supposed to be an internal API, but isn't. Mostly I was just trying to refactor to the point my code compiled and worked before doing a second pass.

The `stream-cipher` crate uses `GenericArray<u8>` here and thus supports type safety around key lengths:

https://docs.rs/stream-cipher/0.3.0/stream_cipher/trait.NewStreamCipher.html#tymethod.new

Reply:
Ah, I see :grin: How is a failure encoded then though? Does it just return a `Self` where the key-length is something invalid?

Reply:
In my bogus trait it panics on an `assert_eq!` below. With `stream-cipher`, all keys that can be expressed as a `GenericArray<u8, Self::KeySize>` at a given key size are valid, so there is never a failure case to deal with.

Reply:
Update: after another round of refactoring, this trait no longer exists, and previous usages have been replaced with the `NewStreamCipher` and `SyncStreamCipher` traits.